datafusion/optimizer/src/type_coercion.rs

```diff
@@ -398,6 +398,20 @@ impl ExprRewriter for TypeCoercionRewriter {
                     }
                 }
             }
+            Expr::ScalarFunction { fun, args } => match fun {
+                BuiltinScalarFunction::Concat
+                | BuiltinScalarFunction::ConcatWithSeparator => {
+                    let new_args = args
```

Comment:
I wonder if we should do something with `LargeUtf8`?

Also, would it make sense to check the types before `clone()`ing them to do a cast that might not be needed?

Reply:
> I wonder if we should do something with LargeUtf8?

This is a good suggestion. My opinion is that we could use `LargeUtf8` if one of the arguments has this type.

Reply:
Is there type coercion rule for the function `Concat` or `ConcatWithSeparator`?

Now the type coercion are not supported in the logical phase for some expr which is `Expr::ScalarFunction`, `Expr::AggregateFunction`,`Expr::WindowFunction` and `Expr::AggregateUDF` in the follow-up pr for this https://github.com/apache/arrow-datafusion/issues/3582#issuecomment-1254441358

Reply:
I think after moving the `type coercion` rule to the logical phase, this issue can be resolved

Reply:
I am not sure what you mean about "type coercion rule for the function Concat or ConcatWithSeparator"

Since it is a `Expr::ScalarFunction { fun, args }` it currently gets coerced using `data_types`  https://github.com/apache/arrow-datafusion/blob/3eb55e9a0510d872f6f7765b1a5f17db46486e45/datafusion/expr/src/type_coercion.rs#L44-L47 

Are you suggesting we move the logic that picks what argument types (in this case string) for concat into `data_types`? (I think this is a good idea, for the record)